search/includes/classes/class-versioning.php

```diff
@@ -745,4 +759,176 @@ public function action__ep_delete_indexable( $object_id, $indexable_slug ) {
 		// Clear the flag to return to normal
 		$this->is_doing_object_delete = false;
 	}
+
+	private function is_self_heal_ongoing() {
+		return 1 === wp_cache_get( self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_KEY, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_GROUP );
+	}
+
+	private function mark_self_heal_ongoing() {
+		wp_cache_set( self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_KEY, 1, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_GROUP, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_TTL );
+	}
+
+	private function mark_self_heal_finished() {
+		wp_cache_delete( self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_KEY, self::INDEX_VERSIONS_SELF_HEAL_LOCK_CACHE_GROUP );
+	}
+
+	/**
+	 * Check if the versions are persisted correctly. If not recreate them.
+	 */
+	public function maybe_self_heal() {
+		if ( $this->is_self_heal_ongoing() ) {
+			return;
+		}
+		$this->mark_self_heal_ongoing();
+
+		if ( $this->is_versioning_valid() ) {
+			$this->mark_self_heal_finished();
+			return;
+		}
+
+		$indicies = $this->get_all_accesible_indicies();
+		if ( is_wp_error( $indicies ) ) {
+			$this->mark_self_heal_finished();
+			return;
+		}
+		$versioning = $this->reconstruct_versioning_option( $indicies );
+
+		update_option( self::INDEX_VERSIONS_OPTION, $versioning );
+
+		$this->mark_self_heal_finished();
+	}
+
+	public function is_versioning_valid() {
+		$all_versions = get_option( self::INDEX_VERSIONS_OPTION, array() );
+
+		if ( ! is_array( $all_versions ) || count( $all_versions ) == 0 ) {
+			return false;
+		}
+
+		foreach ( $all_versions as $versions ) {
+			if ( ! is_array( $versions ) || count( $versions ) == 0 ) {
```

Comment:
Having a count of 0 doesn't _necessarily_ mean that the option is invalid, because index version 1 will get automatically added via `get_versions()` if the array is empty. That was done for backwards compatibility of sites that had indexes set up before versioning was added.

In practice though, this might be ok 🤔 I guess my only worry is that when deployed, this will end up "fixing" a bunch of sites that don't yet have any additional index versions, which is _technically_ unnecessary. I'm not sure how that could be solved though...

Reply:
That I think might be ok. First time this code will run it will reconstruct the defaults and on every other hit it will just use those and not reconstruct anything right?